Revise ZEP0 (#59)
zarr-developers/zeps: active/ZEP0000.md

- **Specification ZEP**

Specification ZEPs deal with the changes related to [Zarr Specifications]. The

> SPEC ZEP (abbreviation of Specification ZEP) introduces changes to the

Reviewer: Unless "SPEC" is an acronym, it should not be all caps.

specification of Zarr. The changes related to specification follow a
strict and thorough review process and should be adopted by everyone in the
Zarr community. Specification ZEPs are of two types. They are:

> - **Core protocol ZEP**

Reviewer: what is the difference between a specification zep and a core protocol zep?

Reviewer: ok I see, a core protocol zep is a subtype of a specification zep. What does the word "protocol" mean here? Are we just using that word to avoid re-using the word "specification"?

Reviewer 2: :+1 to eliminating Protocol and sticking with Specification.

Describes a ZEP which involves changes in the core specification of Zarr. Core

> protocol ZEPs (commonly known as Core ZEPs) are a part of SPEC ZEPs and apply

Reviewer: ```suggestion
protocol ZEPs (hereafter referred to as "Core ZEPs") are a type of Spec ZEPs, and apply
```

and voting from the author of the proposed ZEP, Zarr Steering Council, author
(s) of various Zarr implementations (ZIC), and open-source projects/research
groups using Zarr and the general Zarr community. The general advice is that
everyone should be made aware of changes introduced by core ZEPs. Core protocol
ZEPs require community consensus, and developers or users are typically not
free to ignore them.

- **Protocol extension ZEP**

Describes a ZEP which involves addition to the extension of the core
specification protocol for Zarr. Protocol extension ZEPs (commonly known as
extension ZEPs) are part of Spec ZEPs and apply to Zarr storage extension
specification and various implementations under the Zarr Implementations. The
extension ZEPs are designed so that they are not necessary to be adopted by
every other Zarr implementation. The extension ZEPs may or may not include
involvement from the author of various Zarr implementations and the general
Zarr community. The general idea of having an extension ZEP is to add
functionality, data types and features to accommodate emerging and essential
use cases.

> - **Lean ZEP**

Reviewer 3: Can you provide an example for a Lean ZEP? Where would adding a codec fall?

Reviewer 2: I agree with this comment. In particular, how will an author choose between an Extension ZEP and an Lean ZEP?

Author: Please see: https://github.com/zarr-developers/zeps/pull/59#issuecomment-1989703623.

Describes a ZEP proposing a significantly small and straightforward change to
the Zarr specification. Lean ZEPs are designed for minor modifications that do
not require an extensive review process or community consensus. These proposals
aim to streamline the introduction of small improvements or corrections to the
Zarr specification. Unlike other ZEP types, Lean ZEPs may be introduced through
a simplified process, typically by submitting a GitHub pull request directly.
Lean ZEPs focus on non-controversial changes with minimal impact on existing
implementations, encouraging community input while prioritizing swift
implementation and agility. In some cases, Lean ZEPs may include breaking
changes if there is unanimous approval from all participating stakeholders,

> including the ZIC and the broader Zarr community.

Reviewer: This section is unclear on what a lean ZEP actually is (in terms of procedure), and what contributors have to in the lean ZEP process. 

- Is a lean ZEP literally any small PR against the zarr specs repo? In this case, what is the utility of describing this as a ZEP? This reasoning should be stated explicitly, because it is not obvious. 
- Is a lean ZEP _more_ than a small PR against the zarr specs repo, e.g. a PR + some summary document on a web page? If so, who writes that summary document? Is it extra work for the person submitting the PR? This should also be stated explicitly in this text.

Reviewer: closing this, because the content I'm looking for is later on in the document :)

specification of Zarr. The changes related to specification follow a
strict and thorough review process and should be adopted by everyone in the
Zarr community. Specification ZEPs are of two types. They are:

- **Core protocol ZEP**

<!-- Mention Zarr-Spec devs in line: 85-87, after zarr-developers/governance/#44 is finalised -->

Describes a ZEP which involves changes in the core specification of Zarr. Core
protocol ZEPs (hereafter referred to as "Core ZEPs") are a type of Spec ZEPs,
and apply to [Zarr Specifications] and its various implementations under the
[zarr.dev/implementations]. The core protocol ZEPs should be adopted by every
implementation of Zarr and, in general, the overall Zarr community. Core ZEPs
must go through a thorough review process, including involvement, discussion,
and voting from the author of the proposed ZEP, Zarr Steering Council, author
(s) of various Zarr implementations (ZIC), and open-source projects/research
groups using Zarr and the general Zarr community. The general advice is that
everyone should be made aware of changes introduced by core ZEPs. Core protocol
ZEPs require community consensus, and developers or users are typically not

> free to ignore them. 

Reviewer 2: Given the fact that we have decided not to version the V3 spec (see https://github.com/zarr-developers/zarr-specs/issues/277), it would be helpful to state explicitly whether any core protocol ZEP (i.e. "changes in the core specification of Zarr") will generate a new major version of the spec.

In my opinion, these should go together. Any change major enough to warrant a core protocol ZEP would also trigger a major version increment. The next core protocol ZEP will produce Zarr V4.

Author: Done. Added via https://github.com/zarr-developers/zeps/pull/59/commits/3ae1fc5833cc95220e87370c308b20619f4ab931.

> - **Protocol extension ZEP**

Reviewer 4: This is currently under discussion in https://github.com/zarr-developers/zarr-specs/pull/312 -- I can imagine omitting this section if there are any worries and then adding it back in the future.

Some clarity might be needed on if such a ZEP is *creating* an extension or defining a concrete implementation (in which case a ZEP may not be necessary except for a move to "core").

Reviewer 2: I favor removing Protocol Extension ZEP completely in favor of an extension framework that lies outside the ZEP process (as proposed in https://github.com/zarr-developers/zarr-specs/pull/312)

properly formatted, of high quality, and address initial concerns about the
proposal.

## Submitting a ZEP

### For SPEC ZEPs

The proposal should be submitted as a draft ZEP via a GitHub pull
request to the [zarr-specs] repository.

The PR should contain the narrative document explaining the need, importance,
and use case for the change of the ZEP and should be submitted in the
zarr-specs repository with the name `zep-<n>: <title>.md` under
[/zarr-specs/docs/zeps/draft], where `<n>` is an appropriately assigned
four-digit number. The draft ZEP must use the [ZEP X - Template and
Instructions][template.md] file.

The SPEC ZEPs PR should also contain actual changes to the specification under
[/zarr-specs/docs/specs/...] and mention the assigned four-digit ZEP number

> from the ZEP document.

Reviewer 2: 👏👏👏

This fixes my main complaint about the ZEP process! Thanks!!!

The proposal should be submitted as a draft ZEP via a GitHub pull request to
the [zarr-specs] repository with the name `zep-<n>: <title>.md` where `<n>` is
an appropriately assigned four-digit number (e.g., zep-0000.md). The draft ZEP
must use the [ZEP X - Template and Instructions][template.md] file.

A few points to consider while submitting your ZEP:

- It should sound complete. The ideas must make technical sense.
- The title should accurately describe the content.
- The ZEP’s language (spelling, grammar, sentence structure etc.) and code style should be correct and conformant.

<!-- Mention Zarr-Spec devs in line: 194-198, after zarr-developers/governance/#44 is finalised -->

The [Zarr Steering Council] and the [Zarr Implementations Council] will not
unreasonably deny publication of a ZEP. Reasons for denying ZEP include
duplication of effort, being technically unsound, not providing proper
motivation or addressing backwards compatibility, or not taking care of Zarr
[CODE OF CONDUCT].

## Discussing a ZEP

As soon as the draft ZEP is committed to the [zarr-specs] repository, the author
(s) should use the submitted PR for the ZEP as a central place to discuss and

> review its contents. The ZEP author(s) may create additional issues in the

Reviewer 2: 👍

properly formatted, of high quality, and address initial concerns about the
proposal.

## Submitting a ZEP

### For SPEC ZEPs

The proposal should be submitted as a draft ZEP via a GitHub pull
request to the [zarr-specs] repository.

The PR should contain the narrative document explaining the need, importance,
and use case for the change of the ZEP and should be submitted in the

> zarr-specs repository with the name `zep-<n>: <title>.rst` under

Reviewer 2: I would love to move to markdown.

[/zarr-specs/docs/zeps/draft], where `<n>` is an appropriately assigned
four-digit number. The draft ZEP must use the [ZEP X - Template and
Instructions][template.md] file.

The SPEC ZEPs PR should also contain actual changes to the specification under
[/zarr-specs/docs/specs/...] and mention the assigned four-digit ZEP number
from the ZEP document.

It is also highly recommended that the PR should accompany a complete
implementation or at least an implementation PR in one of the repositories
represented at [zarr.dev/implementations].

### For Lean ZEPs

The proposal should be submitted as a draft ZEP via a GitHub pull
request to the [zarr-specs] repository.

The PR should contain the specific changes to the Zarr specification without
requiring a separate narrative document. As Lean ZEPs involve minor and
non-controversial changes, a narrative document and the use of the template and

> instructions file are not necessary.

Reviewer 2: This sounds like a great way to do ZEPs. Everything in one place! 🎉

The Lean ZEPs PR should be submitted in the [zarr-specs] repository with the
name `<zep-<n>: <title>.md` under [/zarr-specs/docs/specs/...], where `<n>` is

> an appropriately assigned four-digit number.

Reviewer: I am confused -- if a lean ZEP does not require a narrative document, what exactly is the `<zep - <n>: <title>.md` document? I thought a lean ZEP was just a PR against the spec, in which case shouldn't the procedure be "open a PR against the spec"?

Author: Fixed in https://github.com/zarr-developers/zeps/pull/59/commits/81b7ca2c9c554b77c0587ace192f0cbacb255a0b.

Reviewer 4: summarizing last week's discussions including @d-v-b, @MSanKeys963 and me:
 * for the three types of changes:
   - ZEPs
   - "lean ZEP"
   - "just a PR"
 * we talked through some specific examples:
   - add a community codec (i.e. extension) -> just a PR, i.e. "patch"
   - move community codec to core codec -> ZEP (ZIC review), i.e. "minor"
   - choosing a default spelling (US vs UK) -> just a PR *but* might want a ZEP to record a decision
   - v4 -> include the world (ZSC, etc.)
 * we did not find a specific case for "lean ZEP" and instead decided it might be a *technical improvement* (template, bot, code, etc.) which helps expedite generating a ZEP from a simple PR.

    - The author communicates the duration of the Reading Phase when submitting
      the ZEP, typically ranging from a suggested 3-4 weeks with flexibility
      for exceptional circumstances.

3. **Feedback Collection**:
    - Authors anticipate and welcome suggestions from readers to enhance the
      proposal, encouraging general comments to refine definitions, add URLs,
      or improve overall readability.

4. **ZIC Acknowledgment**:
    - ZIC members publicly acknowledge their engagement during the Reading
      Phase, ensuring a comprehensive understanding of community involvement.

<!-- Mention Zarr-Spec devs in line: 315-322, after zarr-developers/governance/#44 is finalised -->

5. **VETO Mechanism**:
    - A VETO option is available to ZIC members if the proposal introduces
      potential breaking changes to existing software implementations, serving
      as a precautionary measure.

Authors may utilize the submitted pull request (PR) for gathering comments,
feedback, acknowledgment, and facilitating engagement from ZIC members during
the Reading Phase.

### Implementation Phase

The Implementation Phase encompasses the timeline during which interested

> adopters actively work on integrating the proposal into their open-source

Reviewer 2: Why should we limit implementations to open-source ones? ARC-GIS supports Zarr, for example.

```suggestion
adopters actively work on integrating the proposal into their
```

the involvement and discussion form the consensus on the ZEP from the author of
various Zarr implementations, open-source projects/research groups using Zarr,
the general Zarr community, and anyone else they think should be included in
the discussions.

## Maintenance

In general, ZEPs are no longer modified after they have reached the `Final`
state. The changes made to the Zarr’s specification in the
[zarr-specs] repository are considered the ultimate reference.

However, finalised ZEPs may undergo minor updates as necessary, with adherence
to semantic versioning principles.

Process ZEPs may be updated over time to reflect changes to development
practices and other details. The precise process followed in these cases will
depend on the nature and purpose of the ZEP being updated.

## ZEP Format

ZEPs are UTF-8 encoded text files using the [reStructuredText] format.
Please see the [ZEP X - Template and Instructions][template.md] file and the
[reStructuredTextPrimer] for more information.

## Header Preamble

```
Author: <list of authors’ real names and email addresses>
Status: < Draft | Active | Accepted | Deferred | Rejected | Withdrawn | Final | Superseded >

> Type: <Specification | Lean | Process | Informational>

Reviewer 4: ```suggestion
Type: <Specification | Process | Informational>
```

## How does a ZEP become accepted?

The Zarr Enhancement Proposal (ZEP) undergoes a systematic process to ensure its
successful adoption by software implementations governed by the Zarr
Implementation Council (ZIC). This process guarantees the stability of the
specification, fosters community consensus, and maintains a high standard of
agreement among maintainers.

### Reading Phase

The Reading Phase, or Request for Comments (RFC) phase, initiates the ZEP
proposal process, marking the introduction of the proposal to the Zarr
community.

> 1. **Stakeholder Engagement**:

Reviewer 3: A reader might ask: Who are the stakeholders and how do I contact them?

Author: Mostly the ZIC and ZSC but also the broader community.
The author is supposed to use the submitted PR to spread the word to everyone, but we also have Google groups set up for notification purposes.

Reviewer 2: :+1: to defining stakeholders clearly.
